From 49e652598211091020cc0b6255075fb60f0e4cc6 Mon Sep 17 00:00:00 2001
From: Finley Xiao <finley.xiao@rock-chips.com>
Date: Fri, 19 Jan 2018 18:25:09 +0800
Subject: [PATCH] soc: rockchip: pvtm: Add support for PX30 SoCs

This adds the necessary data for handling pvtm on the px30.

Change-Id: Ida01e8363f8349c379e10d6a4637eb893591789d
Signed-off-by: Finley Xiao <finley.xiao@rock-chips.com>
---
 .../bindings/soc/rockchip/rockchip-pvtm.txt   |  2 +
 drivers/soc/rockchip/rockchip_pvtm.c          | 45 +++++++++++++++++++
 2 files changed, 47 insertions(+)

diff --git a/Documentation/devicetree/bindings/soc/rockchip/rockchip-pvtm.txt b/Documentation/devicetree/bindings/soc/rockchip/rockchip-pvtm.txt
index 836a01ccc9f5..b560f39432fd 100644
--- a/Documentation/devicetree/bindings/soc/rockchip/rockchip-pvtm.txt
+++ b/Documentation/devicetree/bindings/soc/rockchip/rockchip-pvtm.txt
@@ -7,6 +7,8 @@ performance variance caused by chip process, voltage and temperature.
 
 Required properties:
 - compatible: Should be one of the following.
+ - "rockchip,px30-pvtm" - for PX30 SoCs.
+ - "rockchip,px30-pmu-pvtm" - for PX30 SoCs.
  - "rockchip,rk3288-pvtm" - for RK3288 SoCs.
  - "rockchip,rk3366-pvtm" - for RK3366 SoCs.
  - "rockchip,rk3366-pmu-pvtm" - for RK3366 SoCs.
diff --git a/drivers/soc/rockchip/rockchip_pvtm.c b/drivers/soc/rockchip/rockchip_pvtm.c
index 15f76131d75f..7a8830217f03 100644
--- a/drivers/soc/rockchip/rockchip_pvtm.c
+++ b/drivers/soc/rockchip/rockchip_pvtm.c
@@ -29,6 +29,9 @@
 #include <linux/slab.h>
 #include <linux/soc/rockchip/pvtm.h>
 
+#define PX30_PVTM_CORE		0
+#define PX30_PVTM_PMU		1
+
 #define RK3288_PVTM_CORE	0
 #define RK3288_PVTM_GPU		1
 
@@ -233,6 +236,15 @@ static void rockchip_pvtm_delay(unsigned int delay)
 		udelay(us);
 }
 
+static void px30_pvtm_set_ring_sel(struct rockchip_pvtm *pvtm,
+				   unsigned int sub_ch)
+{
+	unsigned int ch = pvtm->channel->ch;
+
+	regmap_write(pvtm->grf, pvtm->con,
+		     wr_mask_bit(sub_ch, (ch * 0x4 + 0x2), 0x3));
+}
+
 static void rk3399_pvtm_set_ring_sel(struct rockchip_pvtm *pvtm,
 				     unsigned int sub_ch)
 {
@@ -312,6 +324,31 @@ static u32 rockchip_pvtm_get_value(struct rockchip_pvtm *pvtm,
 	return val;
 }
 
+static const struct rockchip_pvtm_channel px30_pvtm_channels[] = {
+	PVTM(PX30_PVTM_CORE, "core", 4, 0, 1, 0x4, 0, 0x4),
+};
+
+static const struct rockchip_pvtm_info px30_pvtm = {
+	.con = 0x80,
+	.sta = 0x88,
+	.num_channels = ARRAY_SIZE(px30_pvtm_channels),
+	.channels = px30_pvtm_channels,
+	.get_value = rockchip_pvtm_get_value,
+	.set_ring_sel = px30_pvtm_set_ring_sel,
+};
+
+static const struct rockchip_pvtm_channel px30_pmupvtm_channels[] = {
+	PVTM(PX30_PVTM_PMU, "pmu", 1, 0, 1, 0x4, 0, 0x4),
+};
+
+static const struct rockchip_pvtm_info px30_pmupvtm = {
+	.con = 0x180,
+	.sta = 0x190,
+	.num_channels = ARRAY_SIZE(px30_pmupvtm_channels),
+	.channels = px30_pmupvtm_channels,
+	.get_value = rockchip_pvtm_get_value,
+};
+
 static const struct rockchip_pvtm_channel rk3288_pvtm_channels[] = {
 	PVTM(RK3288_PVTM_CORE, "core", 1, 0, 1, 0x4, 0, 0x4),
 	PVTM(RK3288_PVTM_GPU, "gpu", 1, 8, 9, 0x8, 1, 0x8),
@@ -379,6 +416,14 @@ static const struct rockchip_pvtm_info rk3399_pmupvtm = {
 };
 
 static const struct of_device_id rockchip_pvtm_match[] = {
+	{
+		.compatible = "rockchip,px30-pvtm",
+		.data = (void *)&px30_pvtm,
+	},
+	{
+		.compatible = "rockchip,px30-pmu-pvtm",
+		.data = (void *)&px30_pmupvtm,
+	},
 	{
 		.compatible = "rockchip,rk3288-pvtm",
 		.data = (void *)&rk3288_pvtm,
-- 
2.35.3

